lib/Sema/CSDiagnostics.cpp

```diff
@@ -2352,3 +2352,100 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
   emitDiagnostic(Member, diag::decl_declared_here, Member->getFullName());
   return true;
 }
+
+bool ImplicitCoercionToAnyWarning::diagnoseAsError() {
+  auto E = getAnchor();
+  auto &TC = getTypeChecker();
+  auto &CS = getConstraintSystem();
+
+  auto locator = CS.getConstraintLocator(E, ConstraintLocator::Member);
```

Comment:
Actually I overlooked this originally but this would be incorrect for anything which is not in form of `a.b`, calls have their own locator, as do unresolved members e.g. `.foo`, and subscripts, you'd have to check each one here to build right locator.

Reply:
Hmm what path element do we use for DeclRefExpr or ParenExpr? I couldn't find them in PathElementKind.

Reply:
I suppose if we have a DeclRefExpr, we could just get the decl directly.

Reply:
DeclRefExpr stores declaration yes, and ParenExpr, Tuple and others don’t have anything but type associated with them, so you’ll have to get a decl from type if it’s a nominal.

Reply:
Hmm for a SubscriptExpr, using SubscriptMember as path element gives me the subscript decl. Example:

```
 func test(_ a : CFString!) {
  let dict = NSMutableDictionary()
  let object = NSObject()
  dict[a] = object
}
```

What I want access to is `a` and not dict.

Reply:
I think locator would help you with that if ‘a’ was converted to Any (parameter type) - getAnchor() would point to its DeclRefExpr (used in the argument position).

Reply:
Hmm it seems like I am going the wrong way. I am checking for each kind of expression which is becoming tedious:

```
if (auto DRE = dyn_cast<DeclRefExpr>(E)) {
	/* Already handled above - we created a locator without a specific path */
} else if (auto UDE = dyn_cast<UnresolvedDotExpr>(E)) {
	locator = CS.getConstraintLocator(E, ConstraintLocator::Member);
} else if (auto SE = dyn_cast<SubscriptExpr>(E)) {
	locator = CS.getConstraintLocator(SE->getIndex()->getValueProvidingExpr());
} else if (auto CE = dyn_cast<CallExpr>(E)) {
	locator = CS.getConstraintLocator(CE->getFn());
} else if (auto MRE = dyn_cast<MemberRefExpr>(E)) {
	locator = CS.getConstraintLocator(MRE->getBase());
} else if (auto AE = dyn_cast<ApplyExpr>(E)) {
	locator = CS.getConstraintLocator(AE->getFn());
} else if (auto PE = dyn_cast<ParenExpr>(E)) {
	locator = CS.getConstraintLocator(PE->getSubExpr());
} else if (auto CCCE = dyn_cast<ConditionalCheckedCastExpr>(E)) {
	locator = CS.getConstraintLocator(CCCE->getSubExpr());
} else if (auto TE = dyn_cast<AnyTryExpr>(E)) {
	locator = CS.getConstraintLocator(TE->getSubExpr());
	  if (isa<CallExpr>(TE->getSubExpr())) {
		auto fn = dyn_cast<CallExpr>(TE->getSubExpr())->getFn();
		locator = CS.getConstraintLocator(fn);
	  }
    }
}
```

Is there a better way to get to the locator that I need to get its decl so I can check for the IUO attribute?

Reply:
I don't think you need to do that `getAnchor()->getSemanticsProvidingExpr()` would already point you to the right argument expression, you just need to handle `DeclRefExpr`, `UnresolvedDotExpr`/`MemberRefExpr`, `UnresolvedMemberExpr` and `OverloadedDeclRefExpr`, that's only expressions which could have declarations associated with them.

Reply:
Hmm maybe not always:

```
func warnOptionalToAnyCoercion(value x: Int?) -> Any {
  do {
    let _: Any = try throwing()
  } catch {}
}

func throwing() throws -> Int? {}
```

Here, the `getSemanticsProvidingExpr` would point to a `CallExpr`, so I guess I need to handle that. But besides that, I also need to figure out how to match these expressions to the right PathElementKind. For example, I can pass `ConstraintLocator::UnresolvedMember` for  `UnresolvedMemberExpr`, but not sure about the rest.

Reply:
Ah yeah and `ApplyExpr` too! For that case you'd have to look at `getFn()` expression. You'd also have to take a look at how locators are formed to generate constraints for the other cases in `CSGen.cpp`, look for `visit*Expr` methods in `ConstraintGenerator` class.

E.g. for `SubscriptExpr` it's going to be `SubscriptMember`, for `UnresolvedDotExpr` and `MemberRefExpr` it's going to be `Member` unless it's a constructor ref than it's `ConstructorMember`.

I think it would be useful to have that as a method in `FailureDiagnostic` just like`getArgumentExprFor`.

Reply:
Thanks for the suggestion @xedin! I have created a new method to return the locator for a given expression - does it look good to you?